Simperium/src/main/java/com/simperium/android/GhostStore.java

```diff
@@ -120,26 +120,21 @@ public void saveGhost(Bucket bucket, Ghost ghost) {
 
     @Override
     public Ghost getGhost(Bucket bucket, String key) throws GhostMissingException {
-        // public Cursor query (String table, String[] columns, String selection, String[] selectionArgs, String groupBy, String having, String orderBy)
         String[] columns = { BUCKET_NAME_FIELD, OBJECT_KEY_FIELD, VERSION_FIELD, PAYLOAD_FIELD };
         String where = "bucketName=? AND simperiumKey=?";
         String[] args = { bucket.getName(), key };
-        Cursor cursor = database.query(GHOSTS_TABLE_NAME, columns, where, args, null, null, null);
-        Ghost ghost = null;
-        if (cursor.getCount() > 0) {
-            cursor.moveToFirst();
-            try {
+
+        try (Cursor cursor = database.query(GHOSTS_TABLE_NAME, columns, where, args, null, null, null)) {
+            if (cursor.moveToFirst()) {
                 JSONObject ghostData = new JSONObject(cursor.getString(3));
-                ghost = new Ghost(cursor.getString(1), cursor.getInt(2), ghostData);
-            } catch (org.json.JSONException e){
-                ghost = null;
+                return new Ghost(cursor.getString(1), cursor.getInt(2), ghostData);
             }
+        } catch (org.json.JSONException e){
+            // a corrupted ghost is effectively equal to a missing ghost
+            // so pass through here and let the library request a new copy
         }
```

Comment:
From what you read in the `try`/`catch` documentation, are there any downsides to calling `cursor.close();` in a `finally` block to ensure the cursor is closed?

Reply:
I used the Java 8 docs to confirm that this should be doing what I think it's doing. Also I made it in a second commit so we could compare with/without.

https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

The relevant bits are in this paragraph

> Prior to Java SE 7, you can use a finally block to ensure that a resource is closed regardless of whether the try statement completes normally or abruptly. The following example uses a finally block instead of a try-with-resources statement:

```java
static String readFirstLineFromFileWithFinallyBlock(String path)
                                                     throws IOException {
    BufferedReader br = new BufferedReader(new FileReader(path));
    try {
        return br.readLine();
    } finally {
        if (br != null) br.close();
    }
}
```

> However, in this example, if the methods readLine and close both throw exceptions, then the method readFirstLineFromFileWithFinallyBlock throws the exception thrown from the finally block; the exception thrown from the try block is suppressed. In contrast, in the example readFirstLineFromFile, if exceptions are thrown from both the try block and the try-with-resources statement, then the method readFirstLineFromFile throws the exception thrown from the try block; the exception thrown from the try-with-resources block is suppressed. In Java SE 7 and later, you can retrieve suppressed exceptions; see the section Suppressed Exceptions for more information.

I don't see `cursor.close()` being able to `throw` here so I don't think there's any benefit or drawback on the runtime. The `try-with-resources` statement though specifically came around for situations like this. I chose it because it took away irrelevant details of this function which made it easier to focus on the logic we're performing.

Reply:
According to the inspection description below, it appears that the `try-with-resources` is effectively the same as the `try-finally` approach.

> Reports try finally statements which can use Java 7 Automatic Resource Management. A quickfix is available to convert the try finally statement into a try with resources statement.
This inspection only reports if the project or module is configured to use a language level of 7.0 or higher.

In that case, `try-with-resources` is fine with me.